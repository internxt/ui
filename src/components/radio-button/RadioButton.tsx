interface RadioButtonProps {
  checked: boolean;
  disabled?: boolean;
  onClick: () => void;
}

export const RadioButton = ({ checked, disabled = false, onClick }: RadioButtonProps) => {
  const borderStyle = disabled ? 'border-gray-10' : 'border-gray-40';
  const checkedStyle =
    disabled && checked ? 'border-0 bg-gray-20' : checked && 'border-0 bg-primary active:bg-primary-dark';

  return (
    <label className="flex cursor-pointer">
      <button
        disabled={disabled}
        onClick={onClick}
        className={`flex h-5 w-5 items-center justify-center rounded-full border ${checkedStyle} ${borderStyle}`}
      >
        {<div className={`h-2.5 w-2.5 rounded-full ${checked ? 'bg-white' : 'hover:bg-gray-10'}`}></div>}
      </button>
      <input type="radio" className="h-0 w-0 appearance-none opacity-0" checked />
    </label>
  );
};
