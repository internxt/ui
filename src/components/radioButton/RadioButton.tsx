export interface RadioButtonProps {
  checked: boolean;
  id?: string;
  disabled?: boolean;
  onClick: (e?: unknown) => void;
}

/**
 * RadioButton component
 *
 * A custom radio button component that allows the user to select one option from a group of choices.
 *
 * @property {boolean} checked
 * - Determines whether the radio button is selected (checked) or not. If true, the radio button appears active.
 *
 * @property {string} [id]
 * - The unique identifier for the radio button element. Useful for associating labels or customizing the radio button.
 *
 * @property {boolean} [disabled]
 * - If true, disables the radio button, making it unclickable and visually indicating its inactive state.
 *
 * @property {(e?: unknown) => void} onClick
 * - A callback function triggered when the radio button is clicked.
 * Can be used to handle custom behavior when the button is selected.
 */

const RadioButton = ({ checked, id, disabled = false, onClick }: RadioButtonProps) => {
  const borderStyle = disabled ? 'border-gray-10' : 'border-gray-40';
  const checkedStyle =
    disabled && checked ? 'border-0 bg-gray-20' : checked && 'border-0 bg-primary active:bg-primary-dark';

  return (
    <div id={id} className="flex cursor-pointer">
      <button
        disabled={disabled}
        onClick={onClick}
        className={`flex h-5 w-5 items-center justify-center rounded-full border ${checkedStyle} ${borderStyle}`}
      >
        {<div className={`h-2.5 w-2.5 rounded-full ${checked || disabled ? 'bg-white' : 'hover:bg-gray-10'}`}></div>}
      </button>
      <input type="radio" className="h-0 w-0 appearance-none opacity-0" checked readOnly />
    </div>
  );
};

export default RadioButton;
