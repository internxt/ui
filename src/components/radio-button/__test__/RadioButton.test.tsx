import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { RadioButton } from '../RadioButton';

describe('Radio Button component', () => {
  it('Radio button onClick should work correctly', () => {
    let checked = false;
    const buttonClick = vi.fn(() => (checked = true));
    render(<RadioButton checked={checked} onClick={buttonClick} />);
    const radioButton = screen.getByRole('button');
    radioButton.click();
    expect(buttonClick).toHaveBeenCalledOnce();
  });

  it('Radio Button component should render correctly', () => {
    const radioButton = render(<RadioButton checked={false} onClick={() => undefined} />);
    expect(radioButton).toMatchSnapshot();
  });

  it('Checked Radio Button component should render correctly', () => {
    const radioButton = render(<RadioButton checked onClick={() => undefined} />);
    expect(radioButton).toMatchSnapshot();
  });

  it('Disabled Radio Button component should render correctly', () => {
    const radioButton = render(<RadioButton disabled checked={false} onClick={() => undefined} />);
    expect(radioButton).toMatchSnapshot();
  });

  it('Disabled and Checked Radio Button component should render correctly', () => {
    const radioButton = render(<RadioButton disabled checked={false} onClick={() => undefined} />);
    expect(radioButton).toMatchSnapshot();
  });
});
