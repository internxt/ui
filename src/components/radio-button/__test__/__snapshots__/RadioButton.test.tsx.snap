// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radio Button component > Checked Radio Button component should render correctly 1`] = `
<div
  className="flex cursor-pointer"
>
  <button
    className="flex h-5 w-5 items-center justify-center rounded-full border border-0 bg-primary active:bg-primary-dark border-gray-40"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="h-2.5 w-2.5 rounded-full bg-white"
    />
  </button>
  <input
    checked={true}
    className="h-0 w-0 appearance-none opacity-0"
    type="radio"
  />
</div>
`;

exports[`Radio Button component > Disabled Radio Button component should render correctly 1`] = `
<div
  className="flex cursor-pointer"
>
  <button
    className="flex h-5 w-5 items-center justify-center rounded-full border false border-gray-10"
    disabled={true}
    onClick={[Function]}
  >
    <div
      className="h-2.5 w-2.5 rounded-full bg-white"
    />
  </button>
  <input
    checked={true}
    className="h-0 w-0 appearance-none opacity-0"
    type="radio"
  />
</div>
`;

exports[`Radio Button component > Disabled and Checked Radio Button component should render correctly 1`] = `
<div
  className="flex cursor-pointer"
>
  <button
    className="flex h-5 w-5 items-center justify-center rounded-full border false border-gray-10"
    disabled={true}
    onClick={[Function]}
  >
    <div
      className="h-2.5 w-2.5 rounded-full bg-white"
    />
  </button>
  <input
    checked={true}
    className="h-0 w-0 appearance-none opacity-0"
    type="radio"
  />
</div>
`;

exports[`Radio Button component > Radio Button component should render correctly 1`] = `
<div
  className="flex cursor-pointer"
>
  <button
    className="flex h-5 w-5 items-center justify-center rounded-full border false border-gray-40"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="h-2.5 w-2.5 rounded-full hover:bg-gray-10"
    />
  </button>
  <input
    checked={true}
    className="h-0 w-0 appearance-none opacity-0"
    type="radio"
  />
</div>
`;
